### PR TITLE
[WFCORE-5239] No suitable provider found for type 'PKCS11' on openjdk…

### PR DESCRIPTION
**Pull request by Skyllarr**
… 11 when configuration is passed in provider-loader

https://issues.redhat.com/browse/WFCORE-5239

### CODE REVIEW

File: elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java [hunk 1]
```diff
@@ -199,12 +202,20 @@ protected boolean dependOnProviderRegistration() {
...
-                                            loadedProviders.add(constructor.newInstance(new Object[] { configSupplier.get() }));
-                                        } catch (NoSuchMethodException ignored) {
-                                            Provider provider = providerClazz.newInstance();
-                                            loadedProviders.add(provider);
-                                            deferred.add(provider::load);
+                                            Constructor<? extends Provider> constructor = doPrivileged((PrivilegedExceptionAction<Constructor<? extends Provider>>) () -> providerClazz.getConstructor(InputStream.class));
+                                            loadedProviders.add(constructor.newInstance(new Object[]{configSupplier.get()}));
+                                        } catch (NoSuchMethodException constructorDoesNotExist) {
+                                            try {
+                                                Method configureMethod = doPrivileged((PrivilegedExceptionAction<Method>) () -> providerClazz.getMethod("configure", String.class));
```
Comment (Skyllarr) [inline elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java:209]: JDK 9 introduced [configure](https://docs.oracle.com/javase/9/docs/api/java/security/Provider.html#configure-java.lang.String-) method to Provider, and the SunPKCS11 constructor where the configuration was passed before was removed. Let me know if usage of this method here could cause backwards incompatibility.

**Comment by Skyllarr** [comment]
@fjuma @darranl This one could use a review. It was created based on a question in wildfly group https://groups.google.com/u/1/g/wildfly/c/mcb_RcJ2AuM/m/ySBMPkmRAgAJ

File: elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java [hunk 2]
```diff
@@ -199,12 +202,20 @@ protected boolean dependOnProviderRegistration() {
...
+                                                Provider provider = providerClazz.newInstance();
+                                                File tempFile = inputStreamToFile(configSupplier);
+                                                loadedProviders.add((Provider) configureMethod.invoke(provider, tempFile.getAbsolutePath()));
+                                                tempFile.delete();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java:213]: It might be good to move the deletion to a `finally` block.

Comment (Skyllarr) [reply elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java]: @fjuma yes thank you. Fixed

File: elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java [hunk 3]
```diff
@@ -236,11 +239,21 @@ protected void resolveRuntime(ModelNode model, OperationContext context) throws
...
+                                                Method configureMethod = doPrivileged((PrivilegedExceptionAction<Method>) () -> providerClazz.getMethod("configure", String.class));
+                                                Provider provider = providerClazz.newInstance();
+                                                tempFile = inputStreamToFile(configSupplier);
+                                                loadedProviders.add((Provider) configureMethod.invoke(provider, tempFile.getAbsolutePath()));
+                                            } catch (NoSuchMethodException configureMethodDoesNotExist) {
+                                                Provider provider = providerClazz.newInstance();
+                                                loadedProviders.add(provider);
+                                                deferred.add(provider::load);
+                                            } finally {
+                                                tempFile.delete();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java:255]: Thanks @Skyllarr! I think just a null check is needed here now.

Comment (Skyllarr) [reply elytron/src/main/java/org/wildfly/extension/elytron/ProviderDefinitions.java]: @fjuma Fixed. Thank you

**Review by fjuma — APPROVED** [review]

**Comment by bstansberry** [comment]
I just stopped CI on this to redirect resources to more time-critical jobs. This one will not be in WF 23 Beta so it's lower priority until that's tagged.

**Review by darranl — APPROVED** [review]

**Pull request merged**